### PR TITLE
add tuplenet api function  [vip:TupleNet-vip#2]

### PR DESCRIPTION
**Pull request by qwjhq**
1、route|static_route|route nat|route_port add del show
    2、switch|switch_port add del show
    3、link switch
    4、edge node init add del(mpa process must run in this edge node)
	new file:   src/control/api/edge.go
	new file:   src/control/api/model.go
	new file:   src/control/api/netRoute.go
	new file:   src/control/api/route.go
	new file:   src/control/api/staticRoute.go
	new file:   src/control/api/switch.go
	new file:   src/control/api/version.go
	modified:   src/control/build.sh
	new file:   src/control/comm/comm.go
	modified:   src/control/go.mod
	modified:   src/control/go.sum
	new file:   src/control/logger/logger.go
	new file:   src/control/tpmpa/README.md
	new file:   src/control/tpmpa/main.go
	new file:   src/control/tpmpa/router.go
	new file:   src/tests/run_tpmpa.sh
	new file:   src/tests/test_simple_tpmpa.sh

### CODE REVIEW

File: src/control/api/edge.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		res.Code = http.StatusBadRequest
+		res.Message = "request phyBr inner virt vip and extGw param"
+		b.Data["json"] = res
+		b.ServeJSON()
+		return
+	}
+
+	endPointArg := "--endpoint=" + etcdHost
+	var preFixArg string
+	if etcdPrefix == defaultEtcdPrefix {
```
Comment (eggqq007) [inline src/control/api/edge.go]: I found duplicate etcdPrefix logic in InitEdge,  AddEdge and DelEdge, please make it become a function. 
opArg , opArg ... those parameters has same problems.

File: src/control/api/edge.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	stdin, _ := cmd.StdinPipe()
+	defer stdin.Close()
+	err := cmd.Start()
+	io.WriteString(stdin, "yes\n")
+
+	if err != nil {
+		logger.Errorf("InitEdge failed phyBr %s inner %s virt %s vip %s extGw %s err %s out %s", phyBr, inner, virt, vip, extGw, cmd.Stderr, cmd.Stdout)
```
Comment (eggqq007) [inline src/control/api/edge.go]: same string in logger.Errorf and res.Message. Please use a var to replace duplicate information.

File: src/control/api/edge.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+		b.Data["json"] = res
+		b.ServeJSON()
+		return
+
+	}
+	err = cmd.Wait()
+	if err != nil {
+		logger.Errorf("InitEdge failed phyBr %s inner %s virt %s vip %s extGw %s err %s out %s", phyBr, inner, virt, vip, extGw, cmd.Stderr, cmd.Stdout)
+		res.Code = http.StatusBadRequest
+		res.Message = fmt.Sprintf("InitAdge failed phyBr %s inner %s virt %s vip %s extGw %s err %s out %s", phyBr, inner, virt, vip, extGw, cmd.Stderr, cmd.Stdout)
```
Comment (eggqq007) [inline src/control/api/edge.go]: ditto

File: src/control/api/model.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	etcdHost      string
+	etcdPrefix    string
+	auth          string
+	edgeShellPath string
+)
+
+// set auth string
+const (
+	defaultAuth          = "YWZhc2Zhc2Zhc2Z3cXJ0cTUxMjVmZ2Znbm82NzgwZmFm"
+	defaultEtcdpoints    = "10.189.114.7:9379"
```
Comment (eggqq007) [inline src/control/api/model.go]: ???

File: src/control/api/model.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	etcdHost = os.Getenv("ETCD_HOSTS")
+	if etcdHost == "" {
+		etcdHost = defaultEtcdpoints
+	}
+	etcdPrefix = os.Getenv("ETCD_PREFIX")
+	if etcdPrefix == "" {
+		etcdPrefix = defaultEtcdPrefix
+	}
+	if controller, err = etcd3.NewController([]string{etcdHost}, etcdPrefix, false); err != nil {
+		logger.Errorf("init connect etcd service failed %s", err)
```
Comment (eggqq007) [inline src/control/api/model.go]: add return?

File: src/control/api/model.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+type NetRequest struct {
+	Route     string `json:route` // route name
+	NatName   string `json:natName`
+	Cidr      string `json:"cidr,omitempty"`
+	XlateType string `json:"xlateType,omitempty"` // snat or dnat
+	XlateIP   string `json:"xlateIP,omitempty"`   // snat or dnat ip
+}
+
+type SwitchRequest struct {
```
Comment (eggqq007) [inline src/control/api/model.go]: SwithcPortRequest

File: src/control/api/route.go [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+// add a logical router (lr)
+func (b *TuplenetAPI) AddRoute() {
+	var (
+		m   RouteRequest
+		res Response
+		err error
+	)
+	// access token check
+	key := b.Ctx.Request.Header.Get("X-TUPLENET-AUTH")
+	if key == "" || key != auth {
```
Comment (eggqq007) [inline src/control/api/route.go]: found same logic in many function. please make a function to check parameter

File: src/control/api/route.go [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+		b.Data["json"] = res
+		b.ServeJSON()
+		return
+	}
+
+	ip, prefix, err := comm.ParseCIDR(cidrString)
+	if err != nil {
+		logger.Errorf("LinkSwitch parse cidr failed route %s switch %s cider string %s", routerName, switchName, cidrString)
+		res.Code = http.StatusInternalServerError
+		res.Message = fmt.Sprintf("LinkSwitch parse cidr failed route %s switch %s cider string %s", routerName, switchName, cidrString)
```
Comment (eggqq007) [inline src/control/api/route.go]: same string in logger and sprintf

File: src/control/api/edge.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	logger.Debugf("AddEdge get param vip %s phyBr %s", vip, phyBr)
+
+	if vip == "" || phyBr == "" {
+		logger.Errorf("AddEdge get param failed vip %s phyBr %s", vip, phyBr)
+		b.BadResponse("request vip and phyBr param")
+		return
+	}
+
+	endPointArg := "--endpoint=" + etcdHost
+	preFixArg := EdgeEtcdPrefix(etcdPrefix)
```
Comment (eggqq007) [inline src/control/api/edge.go]: etcdprefix can be a var, no need recalculation each call

File: src/control/api/switch.go [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+			return
+		}
+	}
+
+	logger.Debugf("DelSwitch success switch %s ", name)
+	b.NormalResponse("DelSwitch success")
+}
+
+// operate on logical switch port(lsp)
+func (b *TuplenetAPI) AddSwitchPort() {
```
Comment (eggqq007) [inline src/control/api/switch.go]: tpctl can add patchport, please add patchport operation

File: src/control/api/staticRoute.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+		srs = []*logicaldev.StaticRoute{r}
+	}
+
+	sort.Slice(srs, func(i, j int) bool { return srs[i].Name < srs[j].Name })
+	logger.Debugf("ShowStaticRoute success route %s rName %s ", name, rName)
+	b.NormalResponse(srs)
+}
+
+func (b *TuplenetAPI) DelStaticRoute() {
```
Comment (eggqq007) [inline src/control/api/staticRoute.go]: each logical entity has show and delete operations. They are similar, please try to use a common function to delete and show them.

File: src/control/api/chassis.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+func (b *TuplenetAPI) DelChassis() {
+	var (
+		m ChassisRequest
+	)
+
+	body, _ := ioutil.ReadAll(b.Ctx.Request.Body)
+	json.Unmarshal(body, &m)
+	name := m.Name
```
Comment (rockmenjack) [inline src/control/api/chassis.go]: consider change to nameOrIP for clarity

File: src/control/api/chassis.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+	err := json.NewDecoder(b.Ctx.Request.Body).Decode(&m)
+	if err != nil {
+		logger.Infof("DelChassis decode body failed %s", err)
+		b.BadResponse("DelChassis decode body failed please check param")
+		return
+	}
+	logger.Infof("DelChassis get param %s", m.NameOrIP)
+
+	if m.NameOrIP == "" {
+		logger.Infof("DelChassis get param failed namOrIP %s", m.NameOrIP)
```
Comment (rockmenjack) [inline src/control/api/chassis.go]: typo: namOrIP

File: src/control/api/chassis.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+		err error
+	)
+	name := b.GetString("name")
+	logger.Infof("ShowChassis get param %s", name)
+
+	if name  == "" { // no name provided show all chassises
+		chs, err = controller.GetChassises()
+		if err != nil {
+			showStr := fmt.Sprintf("ShowChassis get all chassis failed %s ", err)
+			logger.Errorf(showStr)
```
Comment (rockmenjack) [inline src/control/api/chassis.go]: Use logger.Error instead of logger.Errorf

File: src/control/api/edge.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+	io.WriteString(stdin, "yes\n")
+
+	if err != nil {
+		delStr := fmt.Sprintf("DelEdge exec command failed vip %s err %s outErr %s outStr %s", m.Vip, stderr.Bytes(), stdout.Bytes())
+		logger.Errorf(delStr)
+		b.InternalServerErrorResponse(delStr)
+		return
+	}
+
+	err = cmd.Wait()
```
Comment (rockmenjack) [inline src/control/api/edge.go]: add context.Withtimeout

File: src/control/api/edge.go (hunk 3 above)
Comment (qwjhq) [reply src/control/api/edge.go]: done

File: src/control/api/model.go (hunk 6 above)
Comment (qwjhq) [reply src/control/api/model.go]: done

File: src/control/api/route.go (hunk 8 above)
Comment (qwjhq) [reply src/control/api/route.go]: done

File: src/control/api/edge.go (hunk 9 above)
Comment (qwjhq) [reply src/control/api/edge.go]: done

File: src/control/api/switch.go (hunk 10 above)
Comment (qwjhq) [reply src/control/api/switch.go]: done

File: src/control/api/chassis.go (hunk 12 above)
Comment (qwjhq) [reply src/control/api/chassis.go]: done

File: src/control/api/chassis.go (hunk 13 above)
Comment (qwjhq) [reply src/control/api/chassis.go]: done

File: src/control/api/edge.go (hunk 15 above)
Comment (qwjhq) [reply src/control/api/edge.go]: done

File: src/control/api/route.go (hunk 7 above)
Comment (qwjhq) [reply src/control/api/route.go]: done

File: src/control/api/staticRoute.go (hunk 11 above)
Comment (qwjhq) [reply src/control/api/staticRoute.go]: next version will try do it

**Pull request merged**